Changes to add/invoke content provider for broker interaction

kreedula opened the pull request:
MSAL PR : https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1053
Broker PR : https://github.com/AzureAD/ad-accounts-for-android/pull/1298
Change Doc : https://identitydivision.visualstudio.com/DefaultCollection/DevEx/_git/AuthLibrariesApiReview/pullrequest/1554

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationConstants.java:
```diff
@@ -1247,6 +1257,50 @@ private AuthenticationConstants() {
+
+        public static final String REMOVE_ACCOUNTS_PATH = "/removeAccounts";
+
+        public static final String GET_CURRENT_ACCOUNT_SHARED_DEVICe_PATH = "/getCurrentAccountSharedDevice";
```
nit: capitalize letter e

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationConstants.java:
```diff
@@ -1247,6 +1257,50 @@ private AuthenticationConstants() {
+
+        public static final int REMOVE_ACCOUNTS_CODE = 5;
+
+        public static final int GET_CURRENT_ACCOUNT_SHARED_DEVICe_CODE = 6;
```
nit: capitalize letter e

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java:
```diff
@@ -101,8 +101,11 @@ public void disconnect() {
+        final String currentActiveBrokerPackageName = new BrokerValidator(context).
+                getCurrentActiveBrokerPackageName(context);
+
+        if (currentActiveBrokerPackageName == null || currentActiveBrokerPackageName.length() == 0 ||
```
can be simplified to `TextUtils.isEmpty`

shahzaibj commented on common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java:
```diff
@@ -101,8 +101,11 @@ public void disconnect() {
+                getCurrentActiveBrokerPackageName(context);
+
+        if (currentActiveBrokerPackageName == null || currentActiveBrokerPackageName.length() == 0 ||
+                !isMicrosoftAuthServiceSupported(context.getPackageManager(), currentActiveBrokerPackageName)) {
```
Looks like we weren't doing this check before. Why was it NOT needed before? Or why is it needed now?

kreedula replied:
We always did the check, see [here](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/broker/MicrosoftAuthClient.java#L137)

Just refactored here to decouple `active broker` check and this

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/broker/BrokerValidator.java:
```diff
@@ -229,6 +231,25 @@ private X509Certificate getSelfSignedCert(final List<X509Certificate> certs)
+     * @param context
+     * @return String
+     */
+    public String getCurrentActiveBrokerPackageName(@NonNull final Context context) {
```
`@Nullable` return value would be good to document with an annot / javadoc (where presumably `null` indicates no active broker)

iambmelt replied:
Oh, I see this was moved. Still good to have :)

kreedula replied:
updated

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -475,29 +499,44 @@ private Bundle getRequestBundleFromBrokerRequest(@NonNull BrokerRequest brokerRe
     }
 
-    public Bundle getRequestBundleForGetAccounts(@NonNull final CommandParameters parameters) {
+    public Bundle getRequestBundleForGetAccounts(@NonNull final CommandParameters parameters,
```
Just calling out that this change alters binary compatibility of `common`....

kreedula replied:
Ack. these are only used by MSAL for these the request , so we should be good.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -475,29 +499,44 @@ private Bundle getRequestBundleFromBrokerRequest(@NonNull BrokerRequest brokerRe
     }
 
-    public Bundle getRequestBundleForRemoveAccount(@NonNull final RemoveAccountCommandParameters parameters) {
+    public Bundle getRequestBundleForRemoveAccount(@NonNull final RemoveAccountCommandParameters parameters,
```
Same here, changes binary compatibility. Just FYI

kreedula replied:
same as above

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -475,29 +499,44 @@ private Bundle getRequestBundleFromBrokerRequest(@NonNull BrokerRequest brokerRe
     }
 
-    public Bundle getRequestBundleForRemoveAccountFromSharedDevice(@NonNull final RemoveAccountCommandParameters parameters) {
+    public Bundle getRequestBundleForRemoveAccountFromSharedDevice(@NonNull final RemoveAccountCommandParameters parameters,
```
Here too

kreedula replied:
same as above

kreedula commented:
> Left some comments for ACK on binary compat changes, documenting meaning/intent of `null` return value with `getCurrentActiveBrokerPackageName`

Thanks @iambmelt , addressed your comments

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/broker/BrokerValidator.java:
```diff
@@ -229,6 +231,26 @@ private X509Certificate getSelfSignedCert(final List<X509Certificate> certs)
+     * Returns the package that is currently active relative to the Work Account custom account type
+     * Note: either the company portal or the authenticator
+     *
+     * @param context : current active broker package name, null if no broker is available
```
Nit: Should this be the `@return` ? Seems like this content on the wrong attribute

kreedula replied:
Addressed via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/918/commits/477877f2d99d6eb7d752f18d941a682b51665c62

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationConstants.java:
```diff
@@ -1247,6 +1257,50 @@ private AuthenticationConstants() {
     }
 
+
+    public static final class BrokerContentProvider{
```
If this is a constant collection, and it looks like it is, should it have a private constructor?  The other instances of this have documentation describing each constant, should this?  Is there a link to documentation that could be referenced?

kreedula replied:
Addressed via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/918/commits/477877f2d99d6eb7d752f18d941a682b51665c62

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationConstants.java:
```diff
@@ -1247,6 +1257,50 @@ private AuthenticationConstants() {
+        public static final String SIGN_OUT_FROM_SHARED_DEVICE_PATH = "/signOutFromSharedDevice";
+
+
+        /** Codes for URI **/
+        public static final int HELLO_URI_CODE = 1;
+
+        public static final int ACQUIRE_TOKEN_INTERACTIVE_CODE = 2;
+
+        public static final int ACQUIRE_TOKEN_SILENT_CODE = 3;
+
+        public static final int GET_ACCOUNTS_CODE = 4;
+
+        public static final int REMOVE_ACCOUNTS_CODE = 5;
+
+        public static final int GET_CURRENT_ACCOUNT_SHARED_DEVICE_CODE = 6;
+
+        public static final int GET_DEVICE_MODE_CODE = 7;
+
+        public static final int SIGN_OUT_FROM_SHARED_DEVICE_CODE = 8;
+
```
Should these integers be in a separate container?

kreedula replied:
These are related to `BrokerContentProvider` itself , so i'd prefer these to be within these class.

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -415,6 +417,28 @@ public Bundle getRequestBundleForHello(@NonNull final CommandParameters paramete
         return requestBundle;
     }
 
+    public Intent getRequestIntentForAcquireTokenInteractive(@NonNull final Bundle resultBundle,
```
Should there be documentation for this public method?

kreedula replied:
Addressed via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/918/commits/477877f2d99d6eb7d752f18d941a682b51665c62

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -475,29 +499,44 @@ private Bundle getRequestBundleFromBrokerRequest(@NonNull BrokerRequest brokerRe
     }
 
-    public Bundle getRequestBundleForGetAccounts(@NonNull final CommandParameters parameters) {
+    public Bundle getRequestBundleForGetAccounts(@NonNull final CommandParameters parameters,
```
Should we document this public method?

kreedula replied:
Addressed via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/918/commits/477877f2d99d6eb7d752f18d941a682b51665c62

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/internal/request/MsalBrokerRequestAdapter.java:
```diff
@@ -475,29 +499,44 @@ private Bundle getRequestBundleFromBrokerRequest(@NonNull BrokerRequest brokerRe
     }
 
-    public Bundle getRequestBundleForRemoveAccountFromSharedDevice(@NonNull final RemoveAccountCommandParameters parameters) {
+    public Bundle getRequestBundleForRemoveAccountFromSharedDevice(@NonNull final RemoveAccountCommandParameters parameters,
```
Should we document this public method?

kreedula replied:
Addressed via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/918/commits/477877f2d99d6eb7d752f18d941a682b51665c62

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/internal/util/ParcelableUtil.java:
```diff
@@ -0,0 +1,53 @@
+        return bytes;
+    }
+
+    public static Parcel unmarshall(byte[] bytes) {
```
@NonNull?

kreedula replied:
Addressed

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/internal/util/ParcelableUtil.java:
```diff
@@ -0,0 +1,53 @@
+        return parcel;
+    }
+
+    public static <T> T unmarshall(byte[] bytes, Parcelable.Creator<T> creator) {
```
@NonNull?

kreedula replied:
Addressed via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/918/commits/477877f2d99d6eb7d752f18d941a682b51665c62

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/internal/util/ParcelableUtil.java:
```diff
@@ -0,0 +1,53 @@
+
+import androidx.annotation.NonNull;
+
+public class ParcelableUtil {
```
Should we have documentation on this public class?  Should it be final, with a private constructor?  Higher level, is there a reason to prefer static method references over instance methods on a singleton-safe class?

kreedula replied:
Added documentation via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/918/commits/477877f2d99d6eb7d752f18d941a682b51665c62. The intention of this class is have just `util` methods to transform parcelable , hence static methods to keep it simple. I would say it's a coding preference we have been following for all our `Util` classes.

AdamBJohnsonx replied:
I'm fine with this, though it will make it more difficult to induce exceptional behavior in the writeToParcel methods for the purposes of testing.  What I would suggest is that, since it's never reasonable to instantiate or inherit from this type of class, declare it to be final and give it a private constructor.

kreedula replied:
yup, I would not expect to induce any behavior to these as these are basic util methods using Platform API's to transform. Declared as `final` and added a `private` constructor as suggested

AdamBJohnsonx replied:
Thanks!   I'm looking into static analysis to enforce this kind of thing so we're not doing it in review.  Looks like we have a checkstyle ruleset in the codebase...

AdamBJohnsonx commented on common/src/main/java/com/microsoft/identity/common/internal/broker/BrokerValidator.java:
```diff
@@ -229,6 +231,26 @@ private X509Certificate getSelfSignedCert(final List<X509Certificate> certs)
+     * Returns the package that is currently active relative to the Work Account custom account type
+     * Note: either the company portal or the authenticator
+     *
+     * @param context : current active broker package name, null if no broker is available
```
Point of curiosity: what's the relationship between this context and mContext?

kreedula replied:
Good catch, didn't realize we already have a member variable for `Context`, removed `context` as an arg.

kreedula commented:
Thanks @iambmelt and @AdamBJohnsonx  for the reviews. Addressed your comments

kreedula commented on common/src/main/java/com/microsoft/identity/common/internal/util/ParcelableUtil.java:
```diff
@@ -0,0 +1,79 @@
+
+    private ParcelableUtil(){
+
+    }
```
Good tip, now that we have embraced `lombok` , better to use it :)

iambmelt commented:
Taking a look

kreedula commented:
@iambmelt  : Thanks, I was not aware about it. Addressed the comments
